### ISSUE TITLE
Cannot move imported Instagram-pictures to S3

### ISSUE THREAD

**Issue by ghost** [pixelfed/pixelfed#4483]
Hey everyone,

after successfully importing the few Instagram pics I still like today I realized that they aren't moved to my S3-provider (here: Wasabi). Each new pic I upload is nearly instantly stored within S3. When using the "php artisan media:migrate2cloud" to manually push them to S3, the following error appears:

<img width="1026" alt="Bildschirmfoto 2023-06-16 um 14 38 08" src="https://github.com/pixelfed/pixelfed/assets/42076599/ae9172db-d86b-46f6-b563-cdce8b1ff413">

Pixelfed is installed in /var/www/pixelfed and the directory mentioned is also available

![Bildschirmfoto 2023-06-16 um 14 42 45](https://github.com/pixelfed/pixelfed/assets/42076599/a209bd3f-03b7-4a48-9532-1448ff954658)

and linked as everything else is running smoothly in terms of pictures uploaded.

Everything else on there instance works as it should only the imported Instagram-picture are seemingly not uploaded.

The path of a file is, for example, **https://my.pixelfed.tld/storage/m/_v2/555130687540998145/186f813ba-297a89/2saiIi7Tg0Ta/QYb5jKyLg3Ya3B4C0mCZiMGFPHFlKz9bucL5Wdrl.jpg**, seemingly referring to the storage-directory.

Am I missing something or is this maybe just an open issue with the new feature?

Best regards,

Oliver

**Comment by mitexleo** [comment]
This might be a permission issue. Kindly check that.

**Comment by ghost** [comment]
> This might be a permission issue. Kindly check that.

Negative (unfortunately) - the webserver user Pixelfed is running as is inherited throughout the directory and the permissions as stated in [https://docs.pixelfed.org/running-pixelfed/installation/#set-correct-permissions](https://docs.pixelfed.org/running-pixelfed/installation/#set-correct-permissions) are set as well.

**Comment by mitexleo** [comment]
For me  it's unable to create the directories when I ran `php artisan app:transform-imports`. 

Did you managed to run the import job normally ?

**Comment by ghost** [comment]
> For me it's unable to create the directories when I ran `php artisan app:transform-imports`.
> 
> Did you managed to run the import job normally ?

What do you mean? I imported all pics via the option in the backend and they were imported well (apart from this queue-thing which dissolved in the meantime). 😁

`php artisan app:transform-imports`

Just brings... nothing, at least no (visible) output.

As everything with S3 runs without problems it seems to be an issue with imported Insta-pictures only.

**Comment by mitexleo** [comment]
I'm having the same issue I guess 😕

**Comment by mitexleo** [comment]
Did you see any output after running `php artisan app:transform-imports` ?

**Comment by ghost** [comment]
> Did you see any output after running `php artisan app:transform-imports` ?

No, just a line break and no additional output - as if nothing has happened.

**Comment by tsrats** [comment]
I just tried to run `php artisan media:migrate2cloud` now that I have all my instagram photos imported, and while I got the same error "Error migrating media 240 to cloud storage: The file "/opt/webroot/pixelfed/storage/app/" does not exist", it actually did put all my photos onto S3. If I browse in Linode, I can see the entire file structure as it matches on my server, and the images are there, but it looks like the app is not updating the URLs to point to the cloud storage instead of local, nor is it removing the files on local storage after.

**Comment by ghost** [comment]
Apparently (and after doing a backup for security reasons) I used the

`php artisan media:fix-nonlocal-driver`

command on my working instance with the parameters

```
FILESYSTEM_DRIVER=local
FILESYSTEM_CLOUD=s3
```

in my .env-file.

This did indeed take some time but after it was done, the Instagram-import was successfully migrated to S3 and

`php artisan media:migrate2cloud`

didn't throw out errors anymore.

Strange way of achieving this, but this definitely worked... Can anybody confirm?

**Comment by tsrats** [comment]
Interesting, I followed your steps @OliverPifferi, and while it also took a while for `php artisan media:fix-nonlocal-driver` to churn through my photos, when I ran `php artisan media:migrate2cloud`, I still got a lot of the same errors, although it ended with:
 ```
200/200 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100% 
Finished!
Uploaded 92M of media to cloud storage!
 
These files are still stored locally, and will be automatically removed.
```

It did not change the URLs of my photos though, they are still pointing to the local storage :(

EDIT: I spoke too soon, it did start changing the URLs properly now, so I guess that takes time to process?

**Comment by mitexleo** [comment]
I'm still getting a lots of error messages.

**Comment by neonota** [comment]
I'm still getting a lots of error messages.

**Comment by neonota** [comment]
I'm still getting a lots of error messages.

**Comment by mitexleo** [comment]
`200/200 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100% Finished! Uploaded 92M of media to cloud storage! These files are still stored locally, and will be automatically removed`

Don't know why we're getting the same message 😕

**Comment by tsrats** [comment]
Interesting, yeah that is odd. I tried it again and got the same 200/200, but a smaller size. I feel like it might not be accurate haha.
```
 200/200 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
Finished!
Uploaded 86M of media to cloud storage!

These files are still stored locally, and will be automatically removed.
```

**Comment by tsrats** [comment]
To close the loop on this. I yolo'd hard (after making a VM snapshot), and went into the DB to remove all local media that no longer existed (I had an admin user but made a 2nd one to start fresh). Then I ran `php artisan media:fix-nonlocal-driver` followed by `php artisan media:migrate2cloud` over and over until all of my photos got updated to have cdn_urls set. I don't know if I would recommend this unless you are the only user on your instance, have a good backup of it, and get at least a basic understanding of how the DB is laid out though.

**Comment by paulexyz** [comment]
The issue is that there is a time constraint in the select statement for processing and optimizing media. I've removed that alongside another small fix in #4729. The shell executed in the original comment from @OliverPifferi is not needed anymore to upload the media to a remote storage backend since that's usually automatically done by the CatchUnoptimizedMedia job
